File: go-backend/internal/model/credit.go
package model

import (
	"encoding/json"
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

// 积分系统三表（spec task-543）：钱包 / 流水 / 价格。
//
// 数值全部采用厘制整型（BIGINT，单位 0.01 分），"分"换算只发生在 API 边界。
// schema 由 Go 端 GORM AutoMigrate 统一管理，Python 侧只做读写
// （先例：llm_usage.go）。复合唯一索引显式命名 uq_ 前缀，对齐 model.go Namer。

// CreditWallet 用户积分钱包，一人一行；懒创建（EnsureCreditWallet）。
type CreditWallet struct {
	ID         uint  `gorm:"primaryKey;autoIncrement"`
	UserID     uint  `gorm:"uniqueIndex"` // → uq_credit_wallet_user_id
	Balance    int64 // 厘
	TotalSpent int64 // 历史累计净消耗（厘），退款冲减
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

func (CreditWallet) TableName() string { return "credit_wallet" }

// CreditTransaction 积分流水。Type: grant / consume / refund。
// Amount 正=入账、负=出账；BalanceAfter 为记账后余额快照。
// (Source, UserID, BizID) 联合唯一 = 幂等键：重复请求命中唯一约束即已计费，不双扣；
// user_id 入键使跨用户同键互不命中（各扣各的），退款/结算按 (user_id, source, biz_id)
// 精确命中本人流水。
type CreditTransaction struct {
	ID           uint           `gorm:"primaryKey;autoIncrement;index:ix_credit_transaction_user_id_created_at_id,priority:3"`
	UserID       uint           `gorm:"uniqueIndex:uq_credit_transaction_source_user_biz,priority:2;index:ix_credit_transaction_user_id_created_at_id,priority:1"`
	Source       string         `gorm:"size:50;uniqueIndex:uq_credit_transaction_source_user_biz,priority:1"` // translate / nomu_prompt_optimize / design_generate / admin_grant…
	BizID        string         `gorm:"size:64;uniqueIndex:uq_credit_transaction_source_user_biz,priority:3"` // 客户端 Idempotency-Key 或服务端 UUID 兜底
	Type         string         `gorm:"size:20"`                                                              // grant / consume / refund
	Amount       int64          // 厘，正入负出
	BalanceAfter int64          // 厘
	Meta         datatypes.JSON `gorm:"type:jsonb"` // variant、张数等上下文
	CreatedAt    time.Time      `gorm:"index:ix_credit_transaction_user_id_created_at_id,priority:2;default:current_timestamp"`
}

func (CreditTransaction) TableName() string { return "credit_transaction" }

// PreconsumedQty 读 consume 流水 meta.qty（预扣张数）；缺失/非法按 1。
func (t CreditTransaction) PreconsumedQty() int {
	var m struct {
		Qty int `json:"qty"`
	}
	if len(t.Meta) > 0 && json.Unmarshal(t.Meta, &m) == nil && m.Qty > 0 {
		return m.Qty
	}
	return 1
}

// CreditPrice 按次定额价格表。(Source, Variant) 联合唯一；
// 非档位服务 Variant 为空串。调价第一期直改表数据，不做 API。
type CreditPrice struct {
	ID        uint   `gorm:"primaryKey;autoIncrement"`
	Source    string `gorm:"size:50;uniqueIndex:uq_credit_price_source_variant"`
	Variant   string `gorm:"size:20;uniqueIndex:uq_credit_price_source_variant;default:''"`
	UnitPrice int64  `gorm:""` // 厘/次（生图为厘/张）
	CreatedAt time.Time
	UpdatedAt time.Time
}

func (CreditPrice) TableName() string { return "credit_price" }

// creditPriceSeeds 初始定价（厘制 = 展示分 ×100），与 spec 定价表一致。
var creditPriceSeeds = []CreditPrice{
	{Source: "translate", Variant: "", UnitPrice: 10},
	{Source: "nomu_prompt_optimize", Variant: "", UnitPrice: 20},
	{Source: "design_generate", Variant: "lite", UnitPrice: 3000},
	{Source: "design_generate", Variant: "pro", UnitPrice: 4000},
}

// SeedCreditPrices 幂等 seed：冲突（source,variant 已存在）即跳过，不覆盖人工调价。
func SeedCreditPrices(db *gorm.DB) error {
	return db.Clauses(clause.OnConflict{DoNothing: true}).
		CreateInBatches(&creditPriceSeeds, len(creditPriceSeeds)).Error
}

// EnsureCreditWallet 懒创建钱包行（balance=0），并发下靠 user_id 唯一索引兜底，
// 只产生一行。后续扣费在调用方事务内以条件 UPDATE 原子完成。
func EnsureCreditWallet(db *gorm.DB, userID uint) error {
	return db.Clauses(clause.OnConflict{DoNothing: true}).
		Create(&CreditWallet{UserID: userID}).Error
}

File: backend/app/models/credit.py
from __future__ import annotations

from datetime import UTC, datetime

from sqlalchemy import (
    BigInteger,
    DateTime,
    Index,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models import Base


class CreditWallet(Base):
    """用户积分钱包 —— 一人一行，懒创建。

    数值为厘制整型（单位 0.01 分），"分"换算只发生在 API 边界。
    表结构由 Go 端 GORM AutoMigrate 负责（对齐 ``internal/model/credit.go``），
    本模型仅读写；索引名照抄实库（GORM Namer 产物）。
    """

    __tablename__ = "credit_wallet"
    __table_args__ = (
        Index("ix_credit_wallet_userid", "user_id", unique=True),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(BigInteger)
    balance: Mapped[int] = mapped_column(BigInteger, default=0)
    total_spent: Mapped[int] = mapped_column(BigInteger, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC)
    )


class CreditTransaction(Base):
    """积分流水 —— Type 词表: grant / consume / refund。

    ``amount`` 正=入账、负=出账；``balance_after`` 为记账后余额快照。
    ``(source, user_id, biz_id)`` 联合唯一 = 幂等键：重复请求命中唯一约束即
    已计费，不双扣；按 user 隔离，他人同 key 互不命中。表结构由 Go 端 GORM
    AutoMigrate 负责，本模型仅读写（索引名逐字对齐 ``internal/model/credit.go``）。
    """

    __tablename__ = "credit_transaction"
    __table_args__ = (
        UniqueConstraint(
            "source",
            "user_id",
            "biz_id",
            name="uq_credit_transaction_source_user_biz",
        ),
        # user_id 单列查询由下面复合索引的前缀覆盖，与 Go 端一致不另建。
        Index(
            "ix_credit_transaction_user_id_created_at_id",
            "user_id",
            "created_at",
            "id",
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(BigInteger, index=False)
    source: Mapped[str] = mapped_column(String(50))
    biz_id: Mapped[str] = mapped_column(String(64))
    type: Mapped[str] = mapped_column(String(20), index=False)
    amount: Mapped[int] = mapped_column(BigInteger)
    balance_after: Mapped[int] = mapped_column(BigInteger)
    meta: Mapped[dict] = mapped_column(JSONB, default=dict)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        index=False,
    )


class CreditPrice(Base):
    """按次定额价格表 —— ``(source, variant)`` 联合唯一。

    非档位服务 ``variant = ''``（如 translate / nomu_prompt_optimize），
    生图按档定价（design_generate 的 lite / pro）。调价直改表数据，不做 API。
    表结构由 Go 端 GORM AutoMigrate 负责并 seed，本模型只读。
    """

    __tablename__ = "credit_price"
    __table_args__ = (
        UniqueConstraint(
            "source", "variant", name="uq_credit_price_source_variant"
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    source: Mapped[str] = mapped_column(String(50))
    variant: Mapped[str] = mapped_column(
        String(20), default="", server_default=""
    )
    unit_price: Mapped[int] = mapped_column(BigInteger)
    created_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
